### DISCUSSION TITLE
Can't use process.binding directly

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [nodejs/help#4807] [Q&A] [answered]
I'm trying to bypass a JS sandbox and I need to invoke spawn without child_process, so I used the following to test it:

```
let spawn = process.binding('spawn_sync').spawn
let result = spawn('open', ['/System/Applications/Calculator.app'], { shell: true })
console.log(result)
```

However nodejs returns error

```
{ error: -22, status: null, signal: null, output: null, pid: 0 }
```

My questions,

1. I have passed the example same arguments as child_process does, why won't it work?
2. I can't lookup the errors in nodejs source code, I couldn't find that `-22` means, does anyone know?

**Comment by theanarkh** [comment]
parameters of `process.binding('spawn_sync').spawn` are different from `require('childl_process').spawn`, you can try the code as follows.
```
let spawn = process.binding('spawn_sync').spawn
let result = spawn({
    file: '/bin/sh',
    args: ['-c', 'open /System/Applications/Calculator.app'],
    shell: true
})
console.log(result)
```

**Reply by CaledoniaProject** [reply]
@theanarkh With node v16.20.2, still the same error `{ error: -22, status: null, signal: null, output: null, pid: 0 }`

**Comment by JimmyDaddy** [comment, marked as answer]
@CaledoniaProject 

> 1.I have passed the example same arguments as child_process does, why won't it work?

It calls `normalizeSpawnArguments` to format the arguments before calling `spawn_sync.spawn` in `child_process.spawnSync`. [the source code here](https://github.com/nodejs/node/blob/99f6084ef04dd868c7b894ca0fdbb5088773ca6c/lib/child_process.js#L835)

> 2.I can't lookup the errors in nodejs source code, I couldn't find that -22 means, does anyone know?

* `-22` means invalid params
*  `pid = 0` means no process has been started

Maybe you should try the code as follows: 
```js
const spawn = process.binding('spawn_sync').spawn
let result = spawn({
  file: 'open',
  args: ['open', '/System/Applications/Calculator.app'],
  shell: true,
  detached: false,
  stdio: [ // the key is here, you should pass a valid stdio, if stdio is `null` or `undefined` it will returns -22
    { type: 'pipe', readable: true, writable: false }
  ]
})

console.log(result)
```

Hope this helps you.

**Comment by theanarkh** [comment]
Yes, [JimmyDaddy](https://github.com/JimmyDaddy) is right. The parameters look like this.
<img width="463" alt="image" src="https://github.com/nodejs/node/assets/21155906/d125cbc3-2256-442a-a0e2-af0c4a860cfb">

**Reply by CaledoniaProject** [reply]
Awesome, I should have done a debug just like this :-)
